### DISCUSSION TITLE
Information flow from ROS to real UR Robot

### DISCUSSION THREAD

**Discussion by doshininad** [UniversalRobots/Universal_Robots_ROS2_Driver#489] [Q&A]
Can anyone please help me understand, how the trajectory waypoints are send from ROS to the UR controller?
What I mean is: what all information is being transmitted -- waypoints, moveJ, moveL, TCP, APO, Speed, Acceleration, Override etc. ?
All the trajectory waypoints that MoveIt has planned are being send to the robot controller at once or a new waypoint is sent once the old waypoint is reached?

**Comment by fmauch** [comment]
To understand what is happening, you'll first have to understand how ROS's joint trajectory controller (JTC) works. I\ll try to summarize that very briefly.

When a trajectory is being sent to the controller, the controller builds  a spline interpolation based on the trajectory's setpoints. This way, there is a continuous representation of each robot joint w.r.t. joint position, velocity and acceleration. The ROS control loop takes a sample point at the current time from this spline curve which is the current command. Based on the robot's interfaces, this current command will be sent to the robot. This happens in every control loop, so in the case of a UR e-series robot every two milliseconds.
In case of this driver, we use a scaled version of that JTC which respects the speed scaling setting from the robot, let's ignore this fact at the moment, as this doesn't make a huge difference in the question asked.

As you can see from the [controller configuration](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/fabfb3fd63674f0665ba34f8f0780ff9f6e3a9ba/ur_robot_driver/config/ur_controllers.yaml#L85) this driver uses positions being commanded to the robot. Those positions are then being sent to the robot using the `HardwareInterface`. Here, we use what we call the [`ReverseInterface`](https://github.com/UniversalRobots/Universal_Robots_Client_Library#reverseinterface). This is basically a port we open on the ROS machine, where the robot controller connects to. This happens by injecting [script code](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/fabfb3fd63674f0665ba34f8f0780ff9f6e3a9ba/ur_robot_driver/resources/ros_control.urscript) into the robot's program. These position commands are then being interpreted by the script program and pushed into `servoj` commands [here](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/fabfb3fd63674f0665ba34f8f0780ff9f6e3a9ba/ur_robot_driver/resources/ros_control.urscript#L46-L80).

Everything said here could also be implemented using velocity commands, in which case [`speed`](https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/blob/fabfb3fd63674f0665ba34f8f0780ff9f6e3a9ba/ur_robot_driver/resources/ros_control.urscript#L89) would be used. This is currently not setup in the ROS2 driver's controllers.

**Reply by doshininad** [reply]
@fmauch thank you soo much for the detailed explanation.
So, if I am using MoveIt to plan and execute the trajectory. Then, MoveIt will plan the trajectory with some waypoints. These waypoints are then sent to joint trajectory controller which builds the whole trajectory (spline interpolation) and then the ROS control loop sends the current command to the robot at 500Hz frequency. Is this the correct way how it works?

**Reply by fmauch** [reply]
Yes, exactly. In more detail, MoveIt! already does some pre-configuration such that the waypoints sent to the controller are relatively close to each other, each configured with a defined position, velocity and acceleration. This way, the generated spline interpolation will be more or less predefined by those constraints. I'm not 100% familiar with the details of this, but I think that should happen inside the [`trajectory_processing`](https://github.com/ros-planning/moveit2/tree/main/moveit_core/trajectory_processing) package.
